aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -775,23 +781,423 @@ Tensor bmm_nested(const Tensor& self, const Tensor& mat2) {
+    // compute reshaped stride
+    auto opt_stride_reshaped = at::detail::computeStride(size, stride, size_reshaped);
+    // reshape as view is possible
+    if (opt_stride_reshaped.has_value()) {
+      const IntArrayRef& stride_reshaped = *opt_stride_reshaped;
+      // fill reshaped size and stride into sizemat and stridemat
+      for (int64_t idim = 0; idim < ndims_underlying_reshaped; idim++) {
+        sizemat_reshaped_ptr[idim] = size_reshaped[idim];
+        stridemat_reshaped_ptr[idim] = stride_reshaped[idim];
+      }
+      sizemat_reshaped_ptr += ndims_underlying_reshaped;
+      stridemat_reshaped_ptr += ndims_underlying_reshaped;
+    }
+    // reshape as view is impossible
+    else {
+      reshape_as_view = false;
+      // fill reshaped size into sizemat
+      for (int64_t idim = 0; idim < ndims_underlying_reshaped; idim++) {
+        sizemat_reshaped_ptr[idim] = size_reshaped[idim];
+      }
+      sizemat_reshaped_ptr += ndims_underlying_reshaped;
+    }
+  }
+  return std::make_tuple(reshape_as_view, sizemat_reshaped, stridemat_reshaped);
+}
+
+// Args:
+//     nt_reshaped: the reshaped nested tensor to receive copies
+//     buffer: the original nested tensor buffer
+//     sizes: the original nested tensor sizes (may have gone through collapsing or splitting)
+//     strides: the original nested tensor strides (may have gone through collapsing or splitting)
+//     offsets: the original nested tensor offsets (may have gone through collapsing or splitting)
+inline void NestedTensor_reshape_copy(
+    Tensor& nt_reshaped,
+    const Tensor& buffer,
+    const std::vector<IntArrayRef>& sizes,
+    const std::vector<IntArrayRef>& strides,
+    const std::vector<int64_t>& offsets) {
+    auto nt_reshaped_ptr = get_nested_tensor_impl(nt_reshaped);
+    const Tensor& buffer_reshaped = nt_reshaped_ptr->get_buffer();
+    std::vector<IntArrayRef> sizes_reshaped = NestedTensor_get_sizes(nt_reshaped_ptr),
+        strides_reshaped = NestedTensor_get_strides(nt_reshaped_ptr);
+    const std::vector<int64_t>& offsets_reshaped = nt_reshaped_ptr->get_offsets();
+    for (int64_t i = 0; i < nt_reshaped_ptr->size(0); i++) {
+      buffer_reshaped.as_strided(sizes_reshaped[i], strides_reshaped[i], offsets_reshaped[i]).copy_(
+          // TODO: can we avoid allocating new memory for `buffer...reshape`
+          //       I did not find anything like reshape_out
+          buffer.as_strided(sizes[i], strides[i], offsets[i]).reshape(sizes_reshaped[i]));
+    }
+}
+}
+
+// Special rules for reshape(nested tensor):
+// 1. Only 1 regular dimension can be collapsed with
+//    or splitted from the implicit batch dimension
+// 2. Any negative size is acceptable
+// 3. Multiple sizes can be negative
+// 4. Instead of infering size, negative size means "inherit the old size",
+//    so negative size is legal for a ragged dimension.
+Tensor reshape_nested(const Tensor& self, IntArrayRef proposed_shape) {
```

Comment:
I like the detailed comments on the specifics for NT reshape semantics.

We've walked through this code so I'm convinced it's correct, but is it possible to split these cases out into helper functions to make the flow more readable? I'm thinking at least a utility for each case that non-trivially affects the offsets:
* splitting the implicit batch dim
* collapsing the implicit batch dim

It seems like the difference among the 3 cases is mostly limited to offsets changes - is that correct? If so, it should be possible to do something like this rough pseudocode (with better names):
```
if split_case:
    offsets = helper_function_for_split_case(...)
else if collapse_case:
    offsets = helper_function_for_collapse_case(...)

call NestedTensor_reshape_size_stride() and use outputs to do appropriate view / copy
```

Imo it should be possible to pre-calculate everything according to which case we're in then have one place where the actual view / copy reshape is done, if that makes sense

Reply:
A small correction: split and collapse also change size and stride by creating / deleting a leading dimension.

But yes I think we can improve as suggested

Reply:
> A small correction: split and collapse also change size and stride by creating / deleting a leading dimension.

That's right, good point. So my pseudocode above can involve size, stride, and offset determination first then do the meat of it.